docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+
+# Implementation plan
+
+* The sync manager is responsible for persisting the global syncID and version, considering the state to be "new" if it differs, and failing fatally if the version isn't acceptable. Engines never need to see these values.
+
+* The sync manager is responsible for persisting and populating the "declined" engine list. If an engine is declined, the manager should not ask it to sync. IOW, an engine doesn't really need to know whether it is enabled or not. However, engine should expose a way to be "reset" when the manager notices the declined state for an engine has been changed. This will allow engines to optimize certain things, such as not keeping tomstones etc. As part of this, the sync manager must expose an API so the consuming application can change the disabled engine state.
+
+* The sync manager is responsible for managing the "clients" collection - engines should never see details about that engine. However, engines will need the ability to be told about commands targetting that engine. The sync manager will also need to be able to accept commands to be delivered to other clients - although that needs more thought as the conditions in which a command need to be sent typically aren't managed by the engine itself - eg, when a "bookmark" restore is done.
+
+* Each engine is responsible for persisting engine-specific stuff which needs to be persisted. An obvious thing which needs to be persisted is the syncID and version string for the engine itself, so it can take the necessary action when these change.
+
+* The sync manager will manage keys and pass the relevant key into the engine. Keys for engines should not be exposed to engines other than the key it is for (although currently, in practice, all engines share the same key, but this is an implementation detail.
+
+* The sync manager will pass the relevant part of info/collections to the engine and a token which can be used to access storage. Specific engines never talk to the token server.
+
+# Details
+## Structure definitions
+
+While we use rust struct definitions here, it's important to keep in mind that as mentioned above, we'll need to support the manager being written in something other than rust. As the data is relatively small, it may (or may not) be desirable to use JSON as the serialization format.
+
+    // The main structure passed to a sync implementation. This needs to carry
+    // all state necessary to sync.
+    struct GlobalEngineState {
+        // note: no "declined" or "enabled" here - that's invisible to the engine.
+
+        // the result of info/configuration.
+        config: GlobalConfig,
+
+        // The token server token.
+        storage_token: Vec[u8] (??), // not clear on the type here, but whatever.
```

Comment:
µnit: Well, probably needs to be `Vec<u8>` to be Rust syntax at least.

Would it be something like [`Arc<TokenserverToken>`](https://github.com/mozilla/application-services/blob/master/components/sync15/src/token.rs#L30) though? (Answer quite possibly could be no -- I don't know if this is the right token, and even if there is, maybe there's another reason it can't be that...)

Reply:
I left this alone for now (although at face value, a clonable object seems a little better to me)